runtime/interpreter/value.go

```diff
@@ -921,122 +992,126 @@ func (v *ArrayValue) Insert(inter *Interpreter, getLocationRange func() Location
 		})
 	}
 
-	// Use the getter, to make sure lazily decoded values are properly loaded.
-	arrayStaticType := v.StaticType().(ArrayStaticType)
-
-	inter.checkContainerMutation(arrayStaticType.ElementType(), element, getLocationRange)
+	interpreter.checkContainerMutation(v.Type.ElementType(), element, getLocationRange)
 
-	v.modified = true
-
-	element.SetOwner(v.Owner)
-
-	elements := v.Elements()
-
-	//nolint:gocritic
-	v.values = append(
-		elements[:index],
-		append(
-			[]Value{element},
-			elements[index:]...,
-		)...,
+	element = interpreter.TransferValue(
+		getLocationRange,
+		element,
+		nil,
+		v.array.Address(),
 	)
-}
-
-// TODO: unset owner?
-func (v *ArrayValue) Remove(index int, getLocationRange func() LocationRange) Value {
-	v.checkBounds(index, getLocationRange)
-
-	v.modified = true
 
-	elements := v.Elements()
-	result := elements[index]
-
-	lastIndex := len(elements) - 1
-	copy(elements[index:], elements[index+1:])
-
-	// avoid memory leaks by explicitly setting value to nil
-	elements[lastIndex] = nil
-
-	v.values = elements[:lastIndex]
+	err := v.array.Insert(uint64(index), element)
+	if err != nil {
+		panic(ExternalError{err})
```

Comment:
Maybe we can check for `atree.IndexOutOfBoundsError` here instead of checking it earlier.

Reply:
Good idea, refactored in 9dde5eda0997d7d4054bb29cc99eab5fe449f16c